haystack/preview/components/routers/document_joiner.py

```diff
@@ -0,0 +1,142 @@
+import itertools
+import logging
+from collections import defaultdict
+from math import inf
+from typing import List, Optional
+from canals.component.types import Variadic
+
+from haystack.preview import component, Document
+
+
+logger = logging.getLogger(__name__)
+
+
+@component
+class DocumentJoiner:
+    """
+    A component that joins input lists of documents from multiple connections and outputs them as one list.
+
+    The component allows multiple join modes:
+    * concatenate: combine the documents from multiple nodes. Any duplicate documents are discarded.
+                   The score is only determined by the last node that outputs the document.
+    * merge: merge scores of documents from multiple nodes. Optionally, each input score can be given a different
+             `weight` & a `top_k` limit can be set. This mode can also be used for "reranking" retrieved documents.
```

Comment:
What's the result of this join mode, one score..? or do we mean that the score of duplicate documents is merged? If yes, then how about: Merge scores of duplicate Documents coming from multiple components. 

Let's change it to active: Optionally, you can assign a `weight` to the scores and set the `top_k` limit for this join mode. You can also use this join mode to rerank retrieved documents.